pilot/pkg/networking/core/v1alpha3/listener_waypoint.go

```diff
@@ -301,127 +225,120 @@ func (lb *ListenerBuilder) buildWaypointInboundVIP(svcs map[host.Name]*model.Ser
-					HTTP: match.ToChain(httpName),
-				})
-			} else if port.Protocol.IsHTTP() {
-				// Otherwise, just insert HTTP/TCP
-				l.FilterChains = append(l.FilterChains, httpChain)
-			} else {
-				l.FilterChains = append(l.FilterChains, tcpChain)
-			}
-			listeners = append(listeners, l)
+			}},
+		}
+		// Workload IP filtering happens here.
+		ipRange := []*xds.CidrRange{}
+		for _, wlx := range wls {
+			cidr := util.ConvertAddressToCidr(wlx.WorkloadInfo.PodIP)
+			ipRange = append(ipRange, &xds.CidrRange{
+				AddressPrefix: cidr.AddressPrefix,
+				PrefixLen:     cidr.PrefixLen,
+			})
 		}
+		chains = append(chains, directChain)
+		ipMatcher.RangeMatchers = append(ipMatcher.RangeMatchers,
+			&matcher.IPMatcher_IPRangeMatcher{
+				Ranges:  ipRange,
+				OnMatch: match.ToChain(directChain.Name),
+			})
 	}
-	return listeners
+	l := &listener.Listener{
+		Name:              "internal",
+		ListenerSpecifier: &listener.Listener_InternalListener{InternalListener: &listener.Listener_InternalListenerConfig{}},
+		ListenerFilters: []*listener.ListenerFilter{
+			util.InternalListenerSetAddressFilter(),
+			xdsfilters.HTTPInspector,
+		},
+		TrafficDirection: core.TrafficDirection_INBOUND,
+		FilterChains:     chains,
+		FilterChainMatcher: &matcher.Matcher{
+			MatcherType: &matcher.Matcher_MatcherTree_{
+				MatcherTree: &matcher.Matcher_MatcherTree{
+					Input: match.DestinationIP,
+					TreeType: &matcher.Matcher_MatcherTree_CustomMatch{
+						CustomMatch: &xds.TypedExtensionConfig{
+							Name:        "ip",
+							TypedConfig: protoconv.MessageToAny(ipMatcher),
+						},
+					},
+				},
+			},
+		},
+	}
+	return l
 }
 
-func getPorts(services []*model.ServiceInstance) []model.Port {
-	p := map[int]model.Port{}
-	for _, s := range services {
-		p[int(s.Endpoint.EndpointPort)] = model.Port{
-			Port:     int(s.Endpoint.EndpointPort),
-			Protocol: s.ServicePort.Protocol,
-		}
-	}
-	pl := []model.Port{}
-	for _, m := range p {
-		pl = append(pl, m)
+func (lb *ListenerBuilder) buildWaypointInboundOriginateConnect() *listener.Listener {
+	name := "inbound_CONNECT_originate"
+	l := &listener.Listener{
+		Name:              name,
+		UseOriginalDst:    wrappers.Bool(false),
+		ListenerSpecifier: &listener.Listener_InternalListener{InternalListener: &listener.Listener_InternalListenerConfig{}},
+		ListenerFilters:   []*listener.ListenerFilter{util.InternalListenerSetAddressFilter()},
+		FilterChains: []*listener.FilterChain{{
+			Filters: []*listener.Filter{{
+				Name: wellknown.TCPProxy,
+				ConfigType: &listener.Filter_TypedConfig{
+					TypedConfig: protoconv.MessageToAny(&tcp.TcpProxy{
+						StatPrefix:       name,
+						ClusterSpecifier: &tcp.TcpProxy_Cluster{Cluster: name},
+						TunnelingConfig: &tcp.TcpProxy_TunnelingConfig{
+							Hostname: "%DOWNSTREAM_LOCAL_ADDRESS%",
+							HeadersToAdd: []*core.HeaderValueOption{
+								{Header: &core.HeaderValue{Key: "x-envoy-original-dst-host", Value: "%DOWNSTREAM_LOCAL_ADDRESS%"}},
```

Comment:
nit: I see this is just copied, but do we still use this? I had thought we got rid of it at some point. It would be nice if this didn't leak into the request.

Not a blocker at all

Reply:
No idea, I can try without it. This looks like some ztunnel requirement.